File: leetcode/589.n-ary-tree-preorder-traversal/ts/preorder.ts
class _Node {
    val: number
    children: _Node[]

    constructor(val?: number, children?: _Node[]) {
        this.val = (val === undefined ? 0 : val)
        this.children = (children === undefined ? [] : children)
    }
}

const preorder = (root: _Node | null): number[] => {
    const res: number[] = []
    traversal(root, res)

    return res
}

const traversal = (root: _Node | null, res: number[]) => {
    if (root === null) return

    res.push(root.val)

    for (const child of root.children) {
        traversal(child, res)
    }
}

export default preorder

File: leetcode/589.n-ary-tree-preorder-traversal/cpp/preorder.cpp
#include <vector>

class Node
{
public:
    int val;
    std::vector<Node *> children;

    Node() {}

    Node(int _val)
    {
        val = _val;
    }

    Node(int _val, std::vector<Node *> _children)
    {
        val = _val;
        children = _children;
    }
};

class Solution
{
public:
    std::vector<int> preorder(Node *root)
    {
        std::vector<int> res;
        traversal(root, res);

        return res;
    }

private:
    void traversal(Node *root, std::vector<int> &res)
    {
        if (root == nullptr)
            return;

        res.push_back(root->val);

        for (auto n : root->children)
            traversal(n, res);
    }
};
